content/drop-stdlib-forwards-bin-compat.md

```diff
@@ -0,0 +1,295 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: pre-sip
+status: submitted
+title: SIP-NN - Drop Forwards Binary Compatibility of the Scala 2.13 Standard Library
+---
+
+**By: Lukas Rytz**
+
+
+## History
+
+| Date           | Version            |
+|----------------|--------------------|
+| Dec 8, 2022    | Initial Version    |
+
+
+## Summary
+
+I propose to drop the forwards binary compatibility requirement that build tools enforce on the Scala 2.13 standard library.
+This will allow implementing performance optimizations of collection operations that are currently not possible.
+It also unblocks adding new classes and new members to existing classes in the standard library.
+
+
+## Backwards and Forwards Compatibility
+
+A library is backwards binary compatible if code compiled against an old version works with a newer version on the classpath.
+Forwards binary compatibility requires the opposite: code compiled against a new version of a library needs to work with an older version on the classpath.
+A more in-depth explanation of binary compatibility is available on the [Scala documentation site](https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html).
+
+Scala build tools like sbt automatically update dependencies on the classpath to the latest patch version that any other dependency on the classpath requires.
+For example, with the following definition
+
+~~~ scala
+libraryDependencies ++= List(
+  "com.softwaremill.sttp.client3" %% "core" % "3.8.3", // depends on ws 1.3.10
+  "com.softwaremill.sttp.shared"  %% "ws"   % "1.2.7", // for demonstration
+)
+~~~
+
+sbt updates the `ws` library to version 1.3.10.
+Running the `evicted` command in sbt displays all dependencies whose version were changed.
+
+This build tool feature allows library authors to only maintain backwards binary compatibility in new versions, they don't need to maintain forwards binary compatibility.
+Backwards binary compatible changes include the addition of new methods in existing classes and the addition of new classes.
+Such additions don't impact existing code that was compiled against an older version, all definitions that were previously present are still there.
+
+### The Standard Library
+
+The Scala standard library is treated specially by sbt and other build tools, its version is always pinned to the `scalaVersion` of the build definition and never updated automatically.
+
+For example, the `"com.softwaremill.sttp.client3" %% "core" % "3.8.3"` library has a dependency on `"org.scala-lang" % "scala-library" % "2.13.10"` in its POM file.
+When a project uses this version of the sttp client in a project with `scalaVersion` 2.13.8, sbt will put the Scala library version 2.13.8 on the classpath.
```

Comment:
There were several premises that motivated this behavior of sbt, which we can list and reevaluate.

1. In general, sbt assumes that users knows what they are doing enough to specify the **compiling** Scala version. This is encoded as `scalaVersion` key.
2. sbt assumes that the scala-compiler version (`scalaVersion`) must match the `scala-library` version for Scala 2.x.
     1. Before sbt introduced the enforcement, I feel like there were some complaints from the Scala team that they can't reliably predict the scala-library.jar that would be on the classpath during compilation.
     2. As the ultimate variant of the alignment use case, sbt allows users to override `scalaOrganization` to fork `scala-compiler` and `scala-library` together to introduce additional feature to Scala 2.x. See [Typelevel Scala](https://typelevel.org/scala/) and https://github.com/sbt/sbt/pull/2634.